.github/workflows/test.yml

```diff
@@ -102,58 +104,87 @@ jobs:
+    - name: Test multi-arch Docker images locally
       run: |
-        # Enable docker daemon experimental support (for 'pull --platform').
-        config='/etc/docker/daemon.json'
-        if [[ -e "$config" ]]; then
-          sudo sed -i -e 's/{/{ "experimental": true, /' "$config"
-        else
-          echo '{ "experimental": true }' | sudo tee "$config"
-        fi
-        sudo systemctl restart docker
-        printf '%s\n' "#!/bin/sh" "echo 'hello world'" >myscript
         for platform in $DOCKER_PLATFORMS; do
           for ext in '' '-alpine'; do
-            image="${DOCKER_BASE}:${TAG}${ext}"
+            image="localhost:5000/${DOCKER_BASE}:${TAG}${ext}"
             msg="Testing docker image $image on platform $platform"
             line="${msg//?/=}"
             printf "\n${line}\n${msg}\n${line}\n"
             docker pull -q --platform "$platform" "$image"
-            if [ -n "$ext" ]; then
+            if [[ -n "$ext" ]]; then
               echo -n "Image architecture: "
               docker run --rm --entrypoint /bin/sh "$image" -c 'uname -m'
             fi
             version=$(docker run --rm "$image" --version)
             echo "shfmt version: $version"
-            if [ "$TAG" != 'latest' -a "$TAG" != "$version" ]; then
+            if [[ $TAG != "latest" ]] &&
+            [[ $TAG != "$version" ]] &&
+            ! echo "$version" | grep -q "$TAG"; then
               echo "Version mismatch: shfmt $version tagged as $TAG"
               exit 1
             fi
-            docker run --rm -v "$PWD:/mnt" -w '/mnt' "$image" -d myscript
+            docker run --rm -v "$PWD:/mnt" -w '/mnt' "$image" -d cmd/shfmt/docker-entrypoint.sh
           done
         done
+    - name: Check GitHub settings
+      if: >
+        github.event_name == 'push' &&
+        (github.ref == 'refs/heads/master' ||
+        startsWith(github.ref, 'refs/tags/')) &&
+        github.repository == 'mvdan/sh'
+      run: |
+        missing=()
+        [[ -n "${{ secrets.DOCKER_USER }}" ]] || missing+=(DOCKER_USER)
+        [[ -n "${{ secrets.DOCKER_PASSWORD }}" ]] || missing+=(DOCKER_PASSWORD)
+        for i in "${missing[@]}"; do
+          echo "Missing github secret: $i"
+        done
+        (( ${#missing[@]} == 0 )) || exit 1
+        echo "DOCKER_DEPLOY_IMAGES=true" >> $GITHUB_ENV
+    - name: Login to DockerHub
+      if: ${{ env.DOCKER_DEPLOY_IMAGES == 'true' }}
+      uses: docker/login-action@v1
+      with:
+        username: ${{ secrets.DOCKER_USER }}
+        password: ${{ secrets.DOCKER_PASSWORD }}
+    - name: Push images to DockerHub
+      if: ${{ env.DOCKER_DEPLOY_IMAGES == 'true' }}
+      run: |
+        for ext in '' '-alpine'; do
+          image_src="${DOCKER_BASE}:${TAG}${ext}"
+          image_dst="${{ secrets.DOCKER_USER }}/${{ env.DOCKER_REPO }}:${TAG}${ext}"
+          msg="Copy multi-arch docker images to DockerHub ($image)"
+          line="${msg//?/=}"
+          printf "\n${line}\n${msg}\n${line}\n"
```

Comment:
is this printf for debugging purposes? I'm not sure I understand the need for the `//?/=` on msg, for example.

Reply:
This is for formatting the output (it will create a line of the size of `$msg`). I did not knew this parameter expansion but I find id quite cool (this comes from @ArturKlauser I guess).

Reply:
Oh wow :P I think I'm impressed?

Reply:
You're welcome ;)